README.md
# MetaVolcanoR

Gene expression meta-analysis visualization tool.

## Overview

The MetaVolcanoR R package combines differential gene expression results. 
It implements three strategies to summarize gene expression activities from 
different studies. i) Random Effects Model (REM) approach. ii) a naive 
vote-counting approach, and iii) a combining-approach. In all cases, 
MetaVolcano exploits the Volcano plot reasoning to visualize the gene 
expression meta-analysis results.

## Installation
```
devtools::install_github('csbl-usp/MetaVolcanoR')
```

## Usage
Load required libraries.

```
library(MetaVolcanoR,
	data.table,
	dplyr,
	rlang,
	parallel,
	ggplot2,
	cowplot,
	plotly,
	metap,
	metafor,
	topconfects) 
```

### Input Data

Users should provide a named list of data.table/data.frame objects containing 
differential gene expression results. Each object of the list must contain 
*gene name*, *fold change*, and *p-value* variables. It is highly recomended 
to also include *variance* or the confidence intervals of the *fold change* 
variables. 

Take a look at the demo data. It includes differential gene expression results
from five studies. 

```
data(diffexplist)
str(diffexplist)
head(diffexplist[[1]])
```

### Random Effect Model MetaVolcano

The *REM* MetaVolcano summarizes the gene fold change variance of several
studies by a random effect model. Consequently, the *summary p-value* comes
from the meta-analysis model and estimates the probability of the *summary 
fold-change* is equal zero. The *REM* MetaVolcano can highligths the top
*metathr* percentage of consistently perturbed genes. This perturbed ranking
is done by the *topconfects* (https://doi.org/10.1186/s13059-019-1674-7) method.


```
meta_degs_rem <- rem_mv(diffexp=diffexplist,
			pcriteria="pvalue",
			foldchangecol='Log2FC', 
			genenamecol='Symbol',
			geneidcol=NULL,
			collaps=FALSE,
			llcol='CI.L',
			rlcol='CI.R',
			vcol=NULL, 
			cvar=TRUE,
			metathr=0.01,
			jobname="MetaVolcano",
			outputfolder=".", 
			draw='HTML',
			ncores=5)

head(meta_degs_rem@metaresult, 3)

meta_degs_rem@MetaVolcano
```
![REM MetaVolcano](https://github.com/csbl-usp/MetaVolcanoR/blob/master/REM_MV.png)

&nbsp;
The *REM* MetaVolcano can also display the forest plot of a given gene based 
on the REM results.

```
draw_forest(remres=meta_degs_rem,
	    gene="MMP9",
	    genecol="Symbol", 
	    foldchangecol="Log2FC",
	    llcol="CI.L", 
	    rlcol="CI.R",
	    studynames=names(diffexplist),
	    jobname=jobname,
	    outputfolder=outputfolder,
	    draw="HTML")

```
![Forest plot](https://github.com/csbl-usp/MetaVolcanoR/blob/master/MMP9_forestplot.png)


```
draw_forest(remres=meta_degs_rem,
	    gene="COL6A6",
	    genecol="Symbol", 
	    foldchangecol="Log2FC",
	    llcol="CI.L", 
	    rlcol="CI.R",
	    studynames=names(diffexplist),
	    jobname=jobname,
	    outputfolder=outputfolder,
	    draw="HTML")

```
![Forest plot](https://github.com/csbl-usp/MetaVolcanoR/blob/master/COL6A6_forestplot.png)


### Vote-counting approach

MetaVolcano identifies differential expressed genes (DEG) for each study based 
on the user-defined *p-value* and *fold change* thresholds. It displays the 
number of differentially expressed and unperturbed genes per study. In addition,
it plots the inverse cumulative distribution of the consistently DEG so the
user can identify the number of genes whose expression is perturbed in at 
least 1 or n studies.

```
meta_degs_vote <- votecount_mv(diffexp=diffexplist,
			       pcriteria='pvalue',
			       foldchangecol='Log2FC',
			       genenamecol='Symbol',
			       geneidcol=NULL,
			       pvalue=0.05,
			       foldchange=0, 
			       metathr=0.01,
			       collaps=FALSE,
			       jobname="MetaVolcano", 
			       outputfolder=".",
			       draw='HTML',
			       ncores=4)

head(meta_degs_vote@metaresult, 3)

meta_degs_vote@degfreq

```
![DEG by study and cummulative inverse distribution](https://github.com/csbl-usp/MetaVolcanoR/blob/master/deg_by_study.png)

MetaVolcano visualizes genes based on the number of studies where genes were
identified as differentially expressed and the their fold change *sign 
consistency*. It means that a gene that was differentially expressed in five 
studies, from which three of them it was downregulated, will get a *sign 
consistency* score of *2 + (-3) = -1*. Based on the user selection, MetaVolcano
can highligths the top *metathr* percentage of consistently perturbed genes.

```
meta_degs_vote@MetaVolcano
```

![Vote-counting MetaVolcano](https://github.com/csbl-usp/MetaVolcanoR/blob/master/votecounting_MV.png)

### Combining-approach 

The *combinig* MetaVolcano can also summarizes the *fold change* of a gene in
different studies by the *mean* or *median* based on the user preference. In
addition, the *combinig* MetaVolcano summarizes the gene differential
expression *p-values* by mean of the Fisher method. The *combining* MetaVolcano
can highligths the top *metathr* percentage of consistently perturbed genes.


```
meta_degs_comb <- combining_mv(diffexp=diffexplist,
			       pcriteria='pvalue', 
			       foldchangecol='Log2FC',
			       genenamecol='Symbol',
			       geneidcol=NULL,
			       metafc='Mean',
			       metathr=0.01, 
			       collaps=TRUE,
			       jobname="MetaVolcano",
			       outputfolder=".",
			       draw='HTML',
			       ncores=4)

head(meta_degs_comb@metaresult, 3)

meta_degs_comb@MetaVolcano

```
![Combining MetaVolcano](https://github.com/csbl-usp/MetaVolcanoR/blob/master/combining_MV.png)



NEWS
# MetavVolcanoR 0.99.0

* Added a `NEWS.md` file to track changes to the package.


R/rem_mv.R
#' @importFrom parallel mclapply
#' @importFrom topconfects normal_confects
#' @importFrom methods new 'slot<-' show
#' @importFrom plotly as_widget
#' @importFrom htmlwidgets saveWidget 
#' @import dplyr

setOldClass('gg')
setOldClass('ggplot')

#' An S4 class to represent MetaVolcanoR results
#' 
#' @slot input merged diiferential expression inputs \code{data.frame} 
#' @slot inputnames names of the differential expression inputs \code{character}
#' @slot metaresult meta-analysis results \code{data.frame}
#' @slot MetaVolcano plot with meta-analysis results
#' @slot degfreq supplementary figure of the vote-counting MetaVolcano

setClass('MetaVolcano', slots = list(input='data.frame',
				     inputnames='character',
				     metaresult='data.frame',
				     MetaVolcano='gg',
				     degfreq='gg'
				     ))

#' A function to perform the Random Effect Model (REM) MetaVolcano
#'
#' This function runs the 'Random Effect Model' MetaVolcano section
#' @param diffexp list of data.frame/data.table (s) with DE results where lines
#'        are genes
#' @param pcriteria the column name of the pvalue variable <string>
#' @param foldchangecol the column name of the foldchange variable <string>
#' @param genenamecol the column name of the gene name variable <string>
#' @param geneidcol the column name of the gene ID/probe/oligo/transcript
#'        variable <string>
#' @param collaps if probes should be collapsed based on the DE direction
#'        <logical>
#' @param llcol left limit of the fold change coinfidence interval variable
#'        name <string>
#' @param rlcol right limit of the fold change coinfidence interval variable
#'        name <string>
#' @param vcol name of the fold change variance variable <string>
#' @param cvar weather or not to calculate gene variance from confidence 
#'        interval limits <logical>
#' @param metathr top percentage of perturbed genes to be highlighted <double>
#' @param jobname name of the running job <string>
#' @param outputfolder /path where to write the results/
#' @param draw wheather or not to draw the .html visualization <logical>
#' @param ncores the number of processors the user wants to use <integer>
#' @keywords write 'combining meta-analysis' metavolcano
#' @return MetaVolcano object
#' @export
#' @examples
#' data(diffexplist)
#' diffexplist <- lapply(diffexplist, function(del) {
#'     dplyr::filter(del, grepl("MP", Symbol))
#' })
#' mv <- rem_mv(diffexplist, metathr = 0.1)
#' str(mv)
rem_mv <- function(diffexp=list(), pcriteria="pvalue", foldchangecol="Log2FC",
		   genenamecol="Symbol", geneidcol=NULL, collaps=FALSE, 
		   llcol="CI.L", rlcol="CI.R", vcol=NULL, cvar=TRUE, 
		   metathr=0.01, jobname="MetaVolcano", outputfolder=".", 
		   draw='HTML', ncores=1) {

    # ---- Calculating variance from coifidence interval
    if(cvar == TRUE) {
        diffexp <- lapply(diffexp, function(...) calc_vi(..., llcol, rlcol))
    	vcol <- 'vi'
    } else {
    	if(is.null(vcol)) {
	    stop("Oops! If cvar=FALSE, you should provide a variance stimate, 
		  Please, check the vcol parameter.")
	}
    }
    
    if (collaps) {
        # --- Removing non-named genes
        diffexp <- mclapply(diffexp, function(g) {
            g %>%
            dplyr::filter(!!as.name(genenamecol) != "") %>%
            dplyr::filter(!is.na(!!as.name(genenamecol)))
        }, mc.cores = ncores)

        # --- Collapsing redundant geneIDs. Rataining the geneID with the 
        # --- smallest pcriteria
        diffexp <- mclapply(diffexp, function(g) {
            collapse_deg(g, genenamecol, pcriteria)
        }, mc.cores = ncores)

	# --- Subsetting the diffexp inputs
	diffexp <- lapply(diffexp, function(...) dplyr::select(...,
			  dplyr::matches(paste(c(genenamecol, foldchangecol,
			                         llcol, rlcol, vcol), 
					       collapse = '|'))))

        # --- merging DEG results
        diffexp <- rename_col(diffexp, genenamecol, ncores)
        meta_diffexp <- Reduce(function(...) merge(..., by = genenamecol, 
						   all = TRUE), diffexp)
	genecol <- genenamecol

    } else {

	if(is.null(geneidcol)) {
	    geneidcol <- genenamecol
	}

	# Testing if geneIDs are unique
	gid <- vapply(diffexp, function(g) {
                    length(unique(g[[geneidcol]])) == nrow(g)
                }, 
	    logical(1))
        
        if(all(gid)) {
        
	    # --- Subsetting the diffexp inputs
            diffexp <- lapply(diffexp, function(...) dplyr::select(...,
			  dplyr::matches(paste(c(geneidcol, foldchangecol,
						 llcol, rlcol, vcol),
					       collapse = '|'))))

            # --- merging the diffexp inputs	
            diffexp <- rename_col(diffexp, geneidcol, ncores)
            meta_diffexp <- Reduce(function(...) merge(..., 
						       by = geneidcol, 
						       all = TRUE), diffexp)

	    genecol <- geneidcol

	} else {
		
	    stop("the geneidcol contains duplicated values, consider to 
		 set collaps=TRUE")

	}
    }    		
	
    # Calculating the REM summary (metafor)
    meta_diffexp <- cbind(meta_diffexp,
                        do.call(rbind,
                                mclapply(split(meta_diffexp, 
					      meta_diffexp[[genecol]]),
                                        function(g) {
                                            remodel(g, foldchangecol, vcol)
                                        }, mc.cores = ncores)
                        )
    )

    # --- Topconfects ranking
    meta_diffexp <- meta_diffexp %>%
        dplyr::mutate(se = (randomCi.ub - randomCi.lb)/3.92) %>% # 95% conf.int
        dplyr::mutate(index = seq(nrow(meta_diffexp)))

    confects <- normal_confects(meta_diffexp$randomSummary, 
				se=meta_diffexp$se, 
				fdr=0.05, 
				full=TRUE)

    meta_diffexp <- merge(meta_diffexp, 
			  dplyr::select(confects$table, c(index, `rank`)), 
			  by = 'index', all = TRUE)
    
    meta_diffexp <- dplyr::arrange(meta_diffexp, `rank`)
   
    # --- Draw REM MetaVolcano
    gg <- plot_rem(meta_diffexp, jobname, outputfolder, genecol, metathr)

    if(draw == "HTML") {
        
        # --- Writing html device for offline visualization
        saveWidget(as_widget(ggplotly(gg)), 
            paste0(normalizePath(outputfolder), 
	           "/RandomEffectModel_MetaVolcano_", 
	           jobname, ".html"))

    } else if(draw == "PDF") {

        # --- Writing PDF visualization
	pdf(paste0(normalizePath(outputfolder),
	           "/RandomEffectModel_MetaVolcano_", jobname,
	           ".pdf"), width = 7, height = 6)
	     plot(gg)
	dev.off()

    } else {
		
	stop("Seems like you did not provide a right 'draw' parameter. 
	      Try NULL, 'PDF' or 'HTML'")

    }
    
    # Set REM result
    icols <- paste(c(genecol, pcriteria, foldchangecol, llcol, rlcol, vcol), 
		   collapse="|^")
    rcols <- paste(c(genecol, "^random", "^het_", "^error$", "^rank$", 
                   "signcon"), collapse="|")
    result <- new('MetaVolcano', 
		  input=dplyr::select(meta_diffexp, 
				      dplyr::matches(icols)),
		  inputnames=names(diffexp),
		  metaresult=dplyr::select(meta_diffexp,
				       dplyr::matches(rcols)),
		  MetaVolcano=gg,
		  degfreq=ggplot()
		  )
    return(result)
}
